src/censored.jl

```diff
@@ -0,0 +1,234 @@
+"""
+    censored(d::UnivariateDistribution, l::Real, u::Real)
+
+Censor a univariate distribution `d` to the interval `[l, u]`.
+
+The lower bound `l` can be finite or `-Inf` and the upper bound `u` can be finite or
+`Inf`. The function throws an error if `l > u`.
+
+The function falls back to constructing a [`Censored`](@ref) wrapper.
+
+# Implementation
+
+To implement a specialized censored form for distributions of type `D`, the method
+`censored(d::D, l::T, u::T) where {T <: Real}` should be implemented.
+"""
+function censored(d::UnivariateDistribution, l::Real, u::Real)
+    return censored(d, promote(l, u)...)
+end
+censored(d::UnivariateDistribution, l::T, u::T) where {T <: Real} = Censored(d, l, u)
+
+"""
+    Censored
+
+Generic wrapper for a censored distribution.
+"""
+struct Censored{D<:UnivariateDistribution, S<:ValueSupport, T <: Real} <: UnivariateDistribution{S}
+    uncensored::D      # the original distribution (uncensored)
+    lower::T      # lower bound
+    upper::T      # upper bound
+    function Censored(d::UnivariateDistribution, l::T, u::T) where {T <: Real}
+        l ≤ u || error("the lower bound must be less than or equal to the upper bound")
+        new{typeof(d), value_support(typeof(d)), T}(d, l, u)
+    end
+end
+
+params(d::Censored) = tuple(params(d.uncensored)..., d.lower, d.upper)
+partype(d::Censored) = partype(d.uncensored)
+Base.eltype(::Type{<:Censored{D}}) where {D} = eltype(D)
+
+#### Range and Support
+
+islowerbounded(d::Censored) = islowerbounded(d.uncensored) || isfinite(d.lower)
+isupperbounded(d::Censored) = isupperbounded(d.uncensored) || isfinite(d.upper)
+
+minimum(d::Censored) = max(minimum(d.uncensored), d.lower)
+maximum(d::Censored) = min(maximum(d.uncensored), d.upper)
+
+function insupport(d::Censored{<:UnivariateDistribution}, x::Real)
+    return d.lower ≤ x ≤ d.upper && insupport(d.uncensored, x)
+end
+
+#### Show
+
+function show(io::IO, d::Censored)
+    print(io, "Censored(")
+    d0 = d.uncensored
+    uml, namevals = _use_multline_show(d0)
+    uml ? show_multline(io, d0, namevals; newline=false) : show_oneline(io, d0, namevals)
+    print(io, ", range=(", d.lower, ", ", d.upper, "))")
+end
+
+_use_multline_show(d::Censored) = _use_multline_show(d.uncensored)
+
+
+#### Statistics
+
+quantile(d::Censored, p::Real) = clamp(quantile(d.uncensored, p), d.lower, d.upper)
+
+median(d::Censored) = clamp(median(d.uncensored), d.lower, d.upper)
+
+function mean(d::Censored)
```

Comment:
To test these expectations, we'll need a case where `truncated` for a discrete distribution returns a distribution with a known mean, variance, and entropy. Since a truncation of `DiscreteUniform` is also `DiscreteUniform`, that's a simple case we can check; we just need to overload `truncate(::DiscreteUniform, ::Real, ::Real)` similar to the overload we have for `Uniform`.

Reply:
We can compare against the "Mixture" way of constructing the same

Reply:
... but also what you said.

Reply:
Yes, but that also requires a discrete distribution whose truncation has known mean, so we need this either way.

Reply:
Let's add that line to `uniform.jl`